src/Chapter3.hs

```diff
@@ -1007,6 +1157,147 @@ contestants, and write a function that decides the outcome of a fight!
+                 | FirstFighterDefeated
+                 | SecondFighterDefeated
+                 | Draw
+                deriving Show
+
+data Knight = Knight
+    { knightHealth    :: Health
+    , knightAttack    :: Attack
+    , knightDefense   :: Defense
+    } | FightingKnight 
+    { fknightHealth               :: Health
+    , fknightMaxHealth            :: Health
+    , fknightAttack               :: Attack
+    , fknightDefense              :: Defense
+    , fknightRoundsBoostedDefense :: Int
+    }
+    deriving (Show)
+
+noAttack = MkAttack 0
+
+attackKnight :: Knight -> (Knight, Attack)
+attackKnight k = case k of
+    FightingKnight _ _ atk _ _  -> (k, atk)
+    Knight _ atk _              -> (k, atk)
+
+usePotionKnight :: Knight -> (Knight, Attack) 
+usePotionKnight (Knight  h atk def) = (FightingKnight h h atk def 0, noAttack) 
+usePotionKnight (FightingKnight h maxH atk def boostDef) = (newKnight, noAttack)
+      where newHealth = max (unHealth h + 10) (unHealth maxH)
+            newKnight = FightingKnight (MkHealth newHealth) maxH atk def boostDef
+
+castDefSpellKnight :: Knight -> (Knight, Attack) 
+castDefSpellKnight (Knight  h atk def) = (FightingKnight h h atk def 3,  noAttack) 
+castDefSpellKnight (FightingKnight h hMax atk def _) = (FightingKnight h hMax atk def 3, noAttack)
+
+
+data Monster = Monster
+    { monsterHealth    :: Health
+    , monsterAttack   :: Attack
+    }  | FledMonster
+    deriving (Show)
+
+fleeMonster :: Monster -> (Monster, Attack)
+fleeMonster _ = (FledMonster, MkAttack 0)
+
+attackMonster :: Monster -> (Monster,Attack)
+attackMonster m = (m, monsterAttack m)
+
+
+class Fighter a where
+    takeDamage :: a -> Attack -> a
+    defeated   :: a -> Bool
+    fled       :: a -> Bool
+
+instance Fighter Knight where
+    takeDamage :: Knight -> Attack -> Knight
+    takeDamage (Knight  h atk def) a =  takeDamage (FightingKnight h h atk def 0)  a
+    takeDamage (FightingKnight  h maxH atk def boostDef) attack = FightingKnight remainingHealth maxH atk def (max (boostDef-1) 0)
+              where reductionFactor = if boostDef > 0 then 2 else 1
+                    effectiveDmg = max (unAttack attack - (unDefense def) * reductionFactor) 0
+                    remainingHealth = MkHealth (max (unHealth h - effectiveDmg) 0)
+
+
+    defeated :: Knight -> Bool
+    defeated (Knight h _ _) = unHealth h <= 0
+    defeated (FightingKnight h _ _ _ _) = unHealth h <= 0
+
+    fled :: Knight -> Bool
+    fled _ = False
+
+
+instance Fighter Monster where
+    takeDamage :: Monster -> Attack -> Monster
+    takeDamage (Monster h a) (MkAttack atk) = Monster remainingHealth a
+                where remainingHealth = MkHealth (max (unHealth h - atk) 0)
+    
+    defeated :: Monster -> Bool
+    defeated m = unHealth (monsterHealth m) <= 0
+
+        
+    fled :: Monster -> Bool
+    fled FledMonster = True
+    fled _ = False
+
+rotateOne :: [a] -> [a]
+rotateOne l = take (length l) (drop 1 (cycle l))
```

Comment:
That's a correct implementation of this function! Nice work on learning previous lessons :clap: 

Though, in this special case you can implement this function a bit more efficiently. 

```suggestion
rotateOne l = drop 1 l ++ take 1 l
```